### ISSUE TITLE
feat: Gradient accumulation for effective large batch training

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#202]
## Summary

No support for gradient accumulation. For ImageNet with ResNet-50, effective batch sizes of 256+ are standard, but GPU memory may only fit 16–32 samples. Gradient accumulation allows simulating large batches by accumulating gradients over multiple micro-batches before updating weights.

## Current State

```java
// Trainer.java — optimizer.step() called every single batch
for (var batch : trainLoader) {
    model.forward(batch.data(), true);
    loss.compute(predictions, batch.labels());
    model.backward(lossResult.gradient());
    optimizer.step(layers);           // ← updates weights EVERY batch
    model.clearCaches();
}
```

No way to accumulate gradients across multiple forward/backward passes.

### API Design

```java
public final class Trainer {
    private int accumulationSteps = 1;  // default: no accumulation

    /**
     * Accumulate gradients over N micro-batches before updating weights.
     * Effective batch size = batchSize × accumulationSteps.
     *
     * @param steps number of micro-batches to accumulate (1 = no accumulation)
     */
    public Trainer withGradientAccumulation(int steps) {
        this.accumulationSteps = steps;
        return this;
    }
}
```

### Modified Training Loop

```java
int microStep = 0;
for (var batch : trainLoader) {
    var predictions = model.forward(batch.data(), true);
    // Scale loss by accumulation steps for correct averaging
    var lossResult = loss.compute(predictions, batch.labels());
    
    // Scale gradient by 1/accumulationSteps
    var scaledGrad = lossResult.gradient().mul(1.0f / accumulationSteps);
    model.backward(scaledGrad);
    
    microStep++;
    
    if (microStep % accumulationSteps == 0) {
        if (maxGradNorm > 0f) clipGradients();
        optimizer.step(layers);
        model.zeroGradients();     // ← NEW: clear accumulated gradients
        if (scheduler != null) {
            optimizer.setLearningRate(scheduler.lr(globalStep, totalSteps));
        }
        globalStep++;
    }
    
    model.clearCaches();  // always clear forward caches for memory
}
```

### Required Changes

1. **Gradient zeroing**: Need `model.zeroGradients()` or `layer.zeroGradients()` method
   - Currently gradients are overwritten each backward pass
   - With accumulation, backward must **add to** existing gradients
   
2. **Gradient accumulation in backward pass**: 
   - Each layer's `backward()` must accumulate gradients (`+=`) instead of overwriting (`=`)
   - Or: accumulate at the optimizer level by summing gradient tensors

3. **LR scheduler step**: Must step only every `accumulationSteps` micro-batches

### Usage

```java
// Physical batch size: 32, accumulation: 8 → effective batch size: 256
var trainer = new Trainer(model, loss, optimizer)
    .withGradientAccumulation(8)
    .withScheduler(scheduler)
    .withGradientClipping(1.0f);

// DataLoader uses physical batch size
var trainLoader = new DataLoader(dataset, 32, true);

trainer.train(trainLoader, testLoader, TrainingConfig.of(epochs, lr, 32));
```

### Gradient Accumulation Mode in Layers
Currently in `Conv2d.backward()`:
```java
gradKernel = ops.conv2dBackward(...);  // overwrites gradKernel
```

Needs to become:
```java
if (accumulating) {
    gradKernel.addInPlace(newGradKernel);  // accumulate
} else {
    gradKernel = newGradKernel;            // overwrite (current behavior)
}
```

Alternative: Always overwrite in backward, accumulate in a separate buffer at optimizer level.

### Loss Scaling
Gradients must be divided by `accumulationSteps` to maintain correct effective learning rate. Two options:
- Scale loss before backward (simpler, shown above)
- Scale gradients before optimizer step (more flexible)

## Acceptance Criteria

- [ ] `Trainer.withGradientAccumulation(int steps)` enables accumulation
- [ ] `steps=1` preserves exact current behavior (backward compatible)
- [ ] Gradients correctly accumulated across micro-batches
- [ ] Loss/gradients scaled by `1/accumulationSteps` for correct averaging
- [ ] Optimizer step only called every `accumulationSteps` micro-batches
- [ ] LR scheduler steps aligned with optimizer steps
- [ ] Gradient clipping applied to accumulated gradients (not per micro-batch)
- [ ] `model.zeroGradients()` clears all parameter gradients
- [ ] Handles last incomplete accumulation window (end of epoch)
- [ ] Unit test: effective gradient equals sum of micro-batch gradients
- [ ] Unit test: steps=1 produces identical training to current implementation
- [ ] Unit test: LR scheduler steps correctly with accumulation

## References

- [Training ImageNet in 1 Hour (Goyal et al., 2017)](https://arxiv.org/abs/1706.02677) — Linear scaling rule with large batches
- [PyTorch gradient accumulation guide](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation)

**Comment by Samyssmile** [comment]
Implemented in cb425a3. Gradient accumulation via `Trainer.withGradientAccumulation(int steps)` — all layers use zero-then-accumulate pattern (PyTorch-compatible). Also fixes memory leaks in Dense, FeedForward, BatchNorm, and MultiHeadAttention backward passes.
